src/charm.py

```diff
@@ -9,93 +9,377 @@
 
 logger = logging.getLogger(__name__)
 
+PEER = "data-integrator-peers"
+MYSQL = "mysql"
+POSTGRESQL = "postgresql"
+MONGODB = "mongodb"
+KAFKA = "kafka"
+
 
 class IntegratorCharm(CharmBase):
     """Integrator charm that connects to database charms."""
 
     def __init__(self, *args):
         super().__init__(*args)
 
-        self.database = self.model.config.get("database", "")
-
         self.framework.observe(self.on.get_credentials_action, self._on_get_credentials_action)
-        self.framework.observe(self.on.start, self._on_config_changed)
         self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on[PEER].relation_changed, self._on_peer_relation_changed)
 
         # MySQL
-        self.mysql = DatabaseRequires(self, relation_name="mysql", database_name=self.database)
+        self.mysql = DatabaseRequires(
+            self,
+            relation_name=MYSQL,
+            database_name=self.database or "",
+            extra_user_roles=self.extra_user_roles or "",
+        )
         self.framework.observe(self.mysql.on.database_created, self._on_database_created)
+        self.framework.observe(self.on[MYSQL].relation_broken, self._on_relation_broken)
 
         # PostgreSQL
         self.postgresql = DatabaseRequires(
-            self, relation_name="postgresql", database_name=self.database
+            self,
+            relation_name=POSTGRESQL,
+            database_name=self.database or "",
+            extra_user_roles=self.extra_user_roles or "",
         )
         self.framework.observe(self.postgresql.on.database_created, self._on_database_created)
+        self.framework.observe(self.on[POSTGRESQL].relation_broken, self._on_relation_broken)
 
         # MongoDB
-        self.mongodb = DatabaseRequires(self, relation_name="mongodb", database_name=self.database)
+        self.mongodb = DatabaseRequires(
+            self,
+            relation_name=MONGODB,
+            database_name=self.database or "",
+            extra_user_roles=self.extra_user_roles or "",
+        )
         self.framework.observe(self.mongodb.on.database_created, self._on_database_created)
+        self.framework.observe(self.on[MONGODB].relation_broken, self._on_relation_broken)
 
-        self._on_config_changed(None)
+        # Kafka
+        self.kafka = KafkaRequires(
+            self,
+            relation_name=KAFKA,
+            topic=self.topic or "",
+            extra_user_roles=self.extra_user_roles or "",
+        )
+        self.framework.observe(self.kafka.on.topic_created, self._on_topic_created)
+        self.framework.observe(self.on[KAFKA].relation_broken, self._on_relation_broken)
+
+    def _on_relation_broken(self, event: RelationBrokenEvent):
+        """Handle relation broken event."""
+        if not self.unit.is_leader:
+            return
+        # update peer databag to trigger the charm status update
+        self._update_active_relations(event, "broken")
```

Comment:
SUGGESTION

Instead of using strings ("broken", "active", ...), it would be good to use constants or enums. Does the ops framework provide something for this? @marcoppenheimer

Reply:
sort of. I mean, if we use Literal, that helps for the typing only. But what I also meant was to also avoid the use of strings, rather replaced by constants.